docs/design/2290-cadence-ndc.md

```diff
@@ -0,0 +1,310 @@
+
+  `initial version < shared version increment`
+
+When performing failover for one domain from one data center to another data center, the version attched to the domain will be changed by the following rule:
+
+  for all versions which follow `version % (shared version increment) == (active data centers' initial version)`
+  find the smallest version which has `version >= old version in domain`
+
+When there is a data conflict, comparison will be made and workflow history events with the highest version will win.
+
+When a data center is trying to mutate a workflow, version will be checked. A data center can mutate a workflow if and only if
+
+  1. version in the domain belongs to this data center, i.e.
+  `(vesion in domain) % (shared version increment) == (this data centers' initial version)`
+  2. the version of this workflow's last event is not less then version in domain, i.e.
+  `(last event's version) <= (version in domain)`
+
+
+### Version History
+
+Version history is a newly introduced concept which provides high level summary about version information of workflow history.
+
+Whenever there is a new workflow history event generated, the version from domain will be attached. Workflow mutable state will keep track of all hisory events & corresponding version.
+
+Example, version history without data conflict:
+
+T = 0:  adding event with event ID == 1 & version == 1
+```
+| -------- | ------------- | --------------- | ------- |
+| Events                   | Version History           |
+| -------- | ------------- | --------------- | ------- |
+| Event ID | Event Version | Event ID        | Version |
+| -------- | ------------- | --------------- | ------- |
+| 1        | 1             | 1               | 1       |
+| -------- | ------------- | --------------- | ------- |
+```
+
+T = 1:  adding event with event ID == 2 & version == 1
+```
+| -------- | ------------- | --------------- | ------- |
+| Events                   | Version History           |
+| -------- | ------------- | --------------- | ------- |
+| Event ID | Event Version | Event ID        | Version |
+| -------- | ------------- | --------------- | ------- |
+| 1        | 1             | 2               | 1       |
+| 2        | 1             |                 |         |
+| -------- | ------------- | --------------- | ------- |
+```
+
+T = 2:  adding event with event ID == 3 & version == 1
+```
+| -------- | ------------- | --------------- | ------- |
+| Events                   | Version History           |
+| -------- | ------------- | --------------- | ------- |
+| Event ID | Event Version | Event ID        | Version |
+| -------- | ------------- | --------------- | ------- |
+| 1        | 1             | 3               | 1       |
+| 2        | 1             |                 |         |
+| 3        | 1             |                 |         |
+| -------- | ------------- | --------------- | ------- |
+```
+
+T = 3:  domain failover triggered, domain version is now 2
```

Comment:
Does it matter whether this is active or passive? Or are they the same (eventually)?
Maybe worth calling out explicitly.